test/pylib/scylla_server.py

```diff
@@ -0,0 +1,307 @@
+
+    async def rest_api_is_up(self) -> bool:
+        """Test that the Scylla REST API is serving. Can be used as a
+        checker function at start up."""
+        try:
+            async with aiohttp.ClientSession() as s:
+                url = "http://{}:10000/".format(self.cfg["host"])
+                async with s.get(url):
+                    return True
+        except aiohttp.ClientConnectionError:
+            return False
+        # Any other exception may indicate a problem, and is passed to the caller.
+
+    async def start(self) -> None:
+        """Start an installed server. May be used for restarts."""
+        START_TIMEOUT = 300     # seconds
+
+        # Add suite-specific command line options
+        scylla_args = SCYLLA_CMDLINE_OPTIONS + self.cmdline_options
+        env = os.environ.copy()
+        env.clear()     # pass empty env to make user user's SCYLLA_HOME has no impact
+        self.cmd = await asyncio.create_subprocess_exec(
+            self.exe,
+            *scylla_args,
+            cwd=self.cfg["workdir"],
+            stderr=self.log_file,
+            stdout=self.log_file,
+            env=env,
+            preexec_fn=os.setsid,
+        )
+
+        self.start_time = time.time()
+
+        while time.time() < self.start_time + START_TIMEOUT:
+            if self.cmd.returncode:
+                with open(self.log_file_name, 'r') as log_file:
+                    logging.error("failed to start server at host %s", self.cfg["host"])
+                    logging.error("last line of {}:".format(self.log_file_name))
+                    log_file.seek(0, 0)
+                    logging.error(log_file.readlines()[-1].rstrip())
+                    h = logging.getLogger().handlers[0]
+                    logpath = h.baseFilename if hasattr(h, 'baseFilename') else "?"  # type: ignore
+                    raise RuntimeError("""Failed to start server at host {}.
+Check the log files:
+{}
+{}""".format(self.cfg["host"], logpath, self.log_file_name))
+
+            if await self.rest_api_is_up():
+                if await self.cql_is_up():
+                    return
+
+            # Sleep 10 milliseconds and retry
+            await asyncio.sleep(0.1)
+            if self.cfg["seeds"] != self.cfg["host"]:
+                await self.force_schema_migration()
+
+        raise RuntimeError("failed to start server {}, check server log at {}".format(
+            self.host, self.log_file_name))
+
+    async def force_schema_migration(self) -> None:
+        auth = PlainTextAuthProvider(username='cassandra', password='cassandra')
+        with Cluster(contact_points=[self.cfg["seeds"]], auth_provider=auth) as cluster:
+            with cluster.connect() as session:
+                session.execute("CREATE KEYSPACE k WITH REPLICATION = {" +
+                                "'class' : 'SimpleStrategy', 'replication_factor' : 1 }")
+                session.execute("DROP KEYSPACE k")
```

Comment:
Interesting. Does this wait for the schema change to propagate to the entire cluster?

Reply:
By the way, even if it did wait, I don't think this fixes the race between token propagation and streaming.

To fix that you would have to force gossip propagation in the middle of bootstrap, **after** new tokens were inserted by joining nodes but **before** streaming starts. There is no reliable way to do that. Right now you call `force...` after the entire bootstrap procedure has finished, after the "damage" was already done.